### PR TITLE
  Linear Regression: Support for JSON and special characters

### PR DESCRIPTION
**Pull request by hpandeycodeit**
JIRA: MADLIB-1284

  - Modified code in linear.py_in to support special characters and JSON data type
  - Written test case for JSON and Speacial characters
  - Test case checks for JSON and special characters in GPDB 5.x and Special characters in GPDB 4.x

### CODE REVIEW

**Comment by jingyimei** [comment]
@hpandeycodeit Seems couple of dev check test cases failed in Jenkins build. Can you please look into them and fix? Thanks!

**Comment by hpandeycodeit** [comment]
@jingyimei I have fixed the linear regression one. Thanks!

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/721/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/722/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/724/

**Comment by jingyimei** [comment]
Jenkins Ok to test

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/726/

**Comment by jingyimei** [comment]
@hpandeycodeit Can you please rebase this branch from master so that we can do full test? Thanks!

File: src/ports/postgres/modules/regress/linear.py_in [hunk 1]
```diff
@@ -50,6 +84,13 @@ def linregr_train(schema_madlib, source_table, out_table,
                        independent_varname=independent_varname,
                        source_table=source_table))
 
+        # USING Clause does not support expressions. So modifying to the regular
+        # join clause instead.
+        join_clause = " ON " + ' AND '.join("{source_table}.{col} = {temp_lin_rst}.\""
```
Comment (orhankislal) [inline src/ports/postgres/modules/regress/linear.py_in]: I suggest changing this to 
```join_clause = '' if grouping_cols is None else "ON" ...```
Currently the code fails if there are no grouping columns and the `heteroskedasticity_option` is True. We should add a test for this in the dev check as well.

Comment (hpandeycodeit) [reply src/ports/postgres/modules/regress/linear.py_in]: Added the above check and Test case as well.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/732/

File: src/ports/postgres/modules/regress/linear.py_in [hunk 2]
```diff
@@ -134,11 +170,11 @@ def linregr_train(schema_madlib, source_table, out_table,
...
                     , '{source_table}'::varchar           as source_table
                     , '{out_table}'::varchar              as out_table
-                    , '{dependent_varname}'::varchar      as dependent_varname
-                    , '{independent_varname}'::varchar    as independent_varname
+                    , $${dependent_varname}$$::varchar      as dependent_varname
+                    , $${independent_varname}$$::varchar    as independent_varname
                     , {num_rows_processed}::integer       as num_rows_processed
                     , {num_rows_skipped}::integer         as num_missing_rows_skipped
-                    , {grouping_col}::text                as grouping_col
+                    , $${grouping_col}$$::text                as grouping_col
```
[the remark this answers is not in the document]
Comment (orhankislal) [reply src/ports/postgres/modules/regress/linear.py_in]: These still create a problem for modules (like margins) that depend on the model table of linregr. This creates a column with the text `NULL` in it. The old version created a column with no value.

**Pull request closed without merging**